### PR TITLE
O3-3424: Distro Inheritance Tool

### PR DESCRIPTION
**Pull request by wikumChamith**
## Description of what I changed

This allows to inherit distro.properties files. If you want to setup a 3.1.0-SNAPSHOT you could use the following distro.properties file.

```
name=Ref 3.x distro new
distro.artifactId=distro-emr-configuration
distro.groupId=org.openmrs
version=3.1.0-SNAPSHOT

```

You can customize the setup by overriding default properties or produce new ones by adding them to the distro.properties file. Here’s an example that adds a SPA module and updates the version of webservices.rest:

```
name=Ref 3.x distro new
distro.artifactId=distro-emr-configuration
distro.groupId=org.openmrs
version=3.1.0-SNAPSHOT

# Add SPA module
omod.spa=2.0.0-SNAPSHOT

# Override the webservices.rest version to 2.45-SNAPSHOT
omod.webservices.rest=2.45.0-SNAPSHOT
```

## Issue I worked on

see https://openmrs.atlassian.net/browse/O3-3431

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean install` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute the above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java [hunk 1]
```diff
@@ -216,38 +212,26 @@ private DistroProperties resolveDistroProperties(Server server) throws MojoExecu
...
 					platformMode = false;
 					break;
 
 				default:  // distro properties from current directory
-					server.setPlatformVersion(
-							distroProperties.getPlatformVersion(distroHelper, server.getServerTmpDirectory()));
-					server.setVersion(distroProperties.getVersion());
+					String distroArtifactId = distroProperties.getParam("distro.artifactId");
+					String distroGroupId = distroProperties.getParam("distro.groupId");
+					String distroVersion = distroProperties.getParam("version");
```
[the remark this answers is not in the document]
Comment (wikumChamith) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java:222]: Should we use default values if the distro.properties file is missing one or two properties, or should we throw an error instead?

File: sdk-commons/src/main/java/org/openmrs/maven/plugins/utility/DistroHelper.java [hunk 2]
```diff
@@ -488,4 +490,32 @@ public Properties getFrontendProperties(Server server) throws MojoExecutionExcep
...
+                    "https://raw.githubusercontent.com/openmrs/openmrs-distro-referenceapplication/" + server.getVersion() + "/frontend/spa-build-config.json");
+        }
+    }
+
+	public Properties getArtifactProperties(Artifact artifact, Server server, String appShellVersion) throws MojoExecutionException {
+		Properties frontendProperties = getFrontendPropertiesForServer(server);
+		Properties configurationProperties = PropertiesUtils.getConfigurationProperty(artifact);
+		File file = downloadDistro(server.getServerDirectory(), artifact);
+		Properties backendProperties = PropertiesUtils.getDistroProperties(file);
+		Properties spaModuleProperty = PropertiesUtils.getModuleProperty("https://raw.githubusercontent.com/openmrs/openmrs-module-spa/master/pom.xml");
```
[the remark this answers is not in the document]
Comment (wikumChamith) [reply sdk-commons/src/main/java/org/openmrs/maven/plugins/utility/DistroHelper.java]: I updated this to use the version helper.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java [hunk 3]
```diff
@@ -216,38 +212,26 @@ private DistroProperties resolveDistroProperties(Server server) throws MojoExecu
...
+					if(StringUtils.isNotBlank(distroArtifactId) && StringUtils.isNotBlank(distroGroupId) && StringUtils.isNotBlank(distroVersion)) {
```
[the remark this answers is not in the document]
Comment (wikumChamith) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java]: Should we use default values if the distro.properties file is missing one or two properties, or should we throw an error instead?

[the remark this answers is not in the document]
Comment (ibacher) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java]: Hmmm... I was thinking just log a message and continue (so ignoring them altogether), but maybe throwing an error if only some of them are defined is a better experience?

File: sdk-commons/src/main/java/org/openmrs/maven/plugins/utility/DistroHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply sdk-commons/src/main/java/org/openmrs/maven/plugins/utility/DistroHelper.java]: PS I realise this wasn't newly added in the code, just noticed it.

**Comment by ibacher** [comment]
Looks good to me. Can this also be extended to the build-distro command? That's mainly where I think this needs to be applied. The other major feature is to support some type of recursion, i.e., handling if the parent distro has a parent.

**Review by ibacher — APPROVED** [review]

**Comment by wikumChamith** [comment]
> Looks good to me. Can this also be extended to the build-distro command? That's mainly where I think this needs to be applied. The other major feature is to support some type of recursion, i.e., handling if the parent distro has a parent.

Should I create another PR for the build-distro command? I'll include recursive support on this.

**Comment by ibacher** [comment]
Yeah, that's fine. Let's merge this in. Do you have rights to do so?

**Comment by wikumChamith** [comment]
> Yeah, that's fine. Let's merge this in. Do you have rights to do so?

Yes, I can merge the PR.  So should I also create a separate PR for recursive support?

**Pull request merged**